presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -127,14 +132,16 @@
     private static final Pattern TABLE_PARAMETER_SAFE_VALUE_PATTERN = Pattern.compile("^[a-zA-Z0-9]*$");
 
     @Inject
-    public ThriftHiveMetastore(MetastoreLocator metastoreLocator, ThriftHiveMetastoreConfig thriftConfig)
+    public ThriftHiveMetastore(MetastoreLocator metastoreLocator, ThriftHiveMetastoreConfig thriftConfig, HiveConfig hiveConfig, HiveMetastoreAuthentication hiveMetastoreAuthentication)
     {
         this.clientProvider = requireNonNull(metastoreLocator, "metastoreLocator is null");
         this.backoffScaleFactor = thriftConfig.getBackoffScaleFactor();
         this.minBackoffDelay = thriftConfig.getMinBackoffDelay();
         this.maxBackoffDelay = thriftConfig.getMaxBackoffDelay();
         this.maxRetryTime = thriftConfig.getMaxRetryTime();
         this.maxRetries = thriftConfig.getMaxRetries();
+        requireNonNull(hiveMetastoreAuthentication, "hiveMetastoreAuthentication is null");
+        this.impersonationEnabled = thriftConfig.isImpersonationEnabled() && hiveMetastoreAuthentication instanceof KerberosHiveMetastoreAuthentication;
```

Comment:
is this `instanceof` important?

(if it is, we should fail when impersonation is enabled but impossible)

Reply:
This brings up a question I had not previous considered. Why do we need Kerberos to do impersonation? Does the `setUGI()` call fail if not using Kerberos? These two things seem independent.

It looks like impersonation does work without Kerberos, since `StaticMetastoreLocator` will call `setUGI()` if `hive.metastore.username` is configured in `StaticMetastoreConfig`. This is not per-user impersonation, but is still impersonation from Presto-to-metastore.

If Kerberos is required, then we should probably move the `impersonationEnabled` config to `MetastoreKerberosConfig` and add a method `isImpersonationEnabled()` to `HiveMetastoreAuthentication`. The `NoHiveMetastoreAuthentication` implementation would always return `false`, and the Kerberos one would only return `true` if the config is true.

Reply:
I think Kerberos requirement is a left over from the initial implementation which was based on delegation tokens.

Reply:
As talked with @findepi in other issue https://github.com/prestosql/presto/pull/1370#issuecomment-524926283, at least non-kerberized Hadoop doesn't allow to call `set_ugi` twice (https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/TUGIBasedProcessor.java#L146-L149).

This isn't issue unless users set both `hive.metastore.username` and `hive.metastore.thrift.impersonation.enabled`. However, let me remove the instanceof~ block.